tiledb/sm/query/query_condition.h

```diff
@@ -0,0 +1,294 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Defines the QueryCondition class.
+ */
+
+#ifndef TILEDB_QUERY_CONDITION_H
+#define TILEDB_QUERY_CONDITION_H
+
+#include <unordered_set>
+
+#include "tiledb/common/logger.h"
+#include "tiledb/common/status.h"
+#include "tiledb/sm/array_schema/array_schema.h"
+#include "tiledb/sm/enums/query_condition_op.h"
+#include "tiledb/sm/query/result_cell_slab.h"
+#include "tiledb/sm/query/result_tile.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+enum class QueryConditionCombinationOp : uint8_t;
+
+class QueryCondition {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /** Default constructor. */
+  QueryCondition();
+
+  /**
+   * Value constructor.
+   *
+   * @param field_name The name of the field this operation applies to.
+   * @param condition_value The value to compare to.
+   * @param condition_value_size The byte size of condition_value.
+   * @param op The relational operation between the value of the field
+   *     and `condition_value`.
+   */
+  QueryCondition(
+      std::string&& field_name,
+      const void* condition_value,
+      uint64_t condition_value_size,
+      QueryConditionOp op);
+
+  /** Copy constructor. */
+  QueryCondition(const QueryCondition& rhs);
+
+  /** Move constructor. */
+  QueryCondition(QueryCondition&& rhs);
+
+  /** Destructor. */
+  ~QueryCondition();
+
+  /* ********************************* */
+  /*             OPERATORS             */
+  /* ********************************* */
+
+  /** Copy-assignment operator. */
+  QueryCondition& operator=(const QueryCondition& rhs);
+
+  /** Move-assignment operator. */
+  QueryCondition& operator=(QueryCondition&& rhs);
+
+  /* ********************************* */
+  /*                API                */
+  /* ********************************* */
+
+  /**
+   * Verifies that the current state contains supported comparison
+   * operations. Currently, we support the following:
+   *   - Fixed-size, single-value, non-nullable attributes.
```

Comment:
This might be a matter of terminology: this patch supports what I've been calling "fixed-size strings", which you might call "var-sized char".

Reply:
Hmm, that might not be accurate, as `STRING:3` is "fixed-sized" along the same vein as `UINT64:3`, but we don't support it today (as it is a *list* of `CHAR:VAR`, which we will support soon). I would consider any STRING as `CHAR:VAR` internally to TileDB.

Reply:
I've reoriented myself with this patch. This patch supports single-cell fixed attributes for all attributes except for strings. The "fixed string" support is different because the comparison operators are interpreted as lexicographical comparison operators. This is the difference between a "char" attribute and a "string" attribute. We will extend string support to var-size in the future, but the majority of the logic in this path will stay the same. Let's keep this in.